### PR TITLE
Ar mp mark item purchased

### PR DESCRIPTION
**Pull request by arandel1**
## Description

The app now keeps track of how many times a user purchases an item:
- We imported MUI to feature clean, simple checkboxes
- When the user checks the checkbox next to an item, Firebase counts it as a purchase, and the totalPurchase field increments by 1
- When the user checks the checkbox next to an item, Firebase updates the dateLastPurchased field to the time the checkbox was checked.
- After the user checks an item off (checks the checkbox), the box will stay checked for 24 hours, then uncheck itself after 24 hours. Firebase uses the dateLastPurchased timestamp to calculate 24 hours after. 

## Related Issue

closes 9

## Acceptance Criteria

- [X] The ListItem component renders a checkbox with a semantic <label>.
- [X] Checking off the item in the UI also updates the dateLastPurchased and totalPurchases properties on the corresponding Firestore document
- [X] The item is shown as checked for 24 hours after the purchase is made (i.e. we assume the user does not need to buy the item again for at least 1 day). After 24 hours, the item unchecks itself so the user can buy it again.
- [X] The updateItem function in firebase.js has been filled out, and sends updates to the firestore database when an item is checked

## Type of Changes

`feature` `enhancement`

### Before
- No checkboxes
Screen Shot 2024-09-06 at 7 24 37 PM

### After
- User can use checkboxes to mark when they've purchased that item.
- You can see before and after **avocado** is selected: dateLastPurchased updates, and totalPurchases increments up by 1. 
Screen Shot 2024-09-06 at 7 18 19 PM

Screen Shot 2024-09-06 at 7 22 00 PM

Screen Shot 2024-09-06 at 7 18 39 PM

## Testing Steps / QA Criteria

- Pull down from branch
- `npm i` to ensure MUI is installed properly
- open Firestore Database in a separate tab
- Navigate to your unique user code and a list of your choice, and the items inside that list
- Select an item (in my example, I selected **avocado**)
- check the checkbox next to that item
- See `dateLastPurchased` field update to the time you checked the checkbox (or, the current time)
- See the `totalPurchases` increase by 1.

To watch the checkbox uncheck itself in real time: 
- Have browser and Firestore tabs open next to each other
- Manually edit the `dateLastPurchased` field by clicking the pencil next to the field: Edit the field to the day + one minute before the current time (You can mess around with this!)

### CODE REVIEW

File: src/components/ListItem.jsx [hunk 1]
```diff
@@ -1,5 +1,72 @@
...
+			if (checked) {
+				// console.log('Checking if 24 hours passed since last purchase');
+				const is24HoursPassed = has24HoursPassed(purchaseDate);
+				// console.log(`Has 24 hours passed: ${is24HoursPassed}`);
+
+				// If 24 hours have passed, uncheck the checkbox associated with the item.
+				if (is24HoursPassed) {
+					// console.log('Unchecking item because 24 hours have passed');
+					setChecked(false);
+					localStorage.setItem(`checked-${item.name}`, JSON.stringify(false));
```
Comment (sar-mko) [inline src/components/ListItem.jsx:34]: **nitpick ( non-blocking ):** Maybe passing the checked state into since its updating anyway.

**Review by sar-mko — COMMENTED** [review]
Really cool implementation :) Was able to check and uncheck items and see properties in the database, and saw the update on click.

Though when testing the 24 hours feature (In firebase, I changed `dateLastPurchased` to the previous day and a minute after), the page updated after a refresh, but I noticed `isChecked` in the database stayed equal to `true`. It could be how I tested it, so I'm curious if anyone also had this happen. It does update in LocalStorage as well!

**Review by mentalcaries — COMMENTED** [review]
Really cool implementation so far.

What I have noticed though is that I am unable to uncheck the item and recheck it right after. The only issue with that is that the item count gets incremented every single time (so even if it was an accidental check, it still counts as one)

File: src/components/ListItem.jsx [hunk 2]
```diff
@@ -1,5 +1,69 @@
...
+	}, [dateLastPurchased, name, checked]);
+
+	// Store the checked state in localStorage whenever it changes
+	useEffect(() => {
+		localStorage.setItem(`checked-${name}`, JSON.stringify(checked));
+	}, [checked, name]);
+
+	// Handle checkbox change (when user clicks on it)
+	const handleChange = () => {
+		const newCheckedState = !checked;
```
Comment (mentalcaries) [inline src/components/ListItem.jsx:47]: **question:** should the state of the item being checked depend only on the checked state in the app/firestore or should that also be dependent on the date?

File: src/components/ListItem.jsx [hunk 3]
```diff
@@ -1,5 +1,69 @@
...
+	const [checked, setChecked] = useState(() => {
+		const storedChecked = localStorage.getItem(`checked-${name}`);
+		return storedChecked ? JSON.parse(storedChecked) : false;
+	});
+
+	// Function to check if 24 hours have passed
+	const has24HoursPassed = (dateLastPurchased) => {
+		const purchaseTime = dateLastPurchased.getTime(); // Time in milliseconds
+		const currentTime = new Date().getTime(); // Current time in milliseconds
+		const oneDayInMs = 24 * 60 * 60 * 1000; // 24 hours in milliseconds
```
Comment (mentalcaries) [inline src/components/ListItem.jsx:20]: **nit (non-blocking):** as this is a constant, consider renaming in uppercase snake case. Note: this constant exists in `dates.js`

File: src/api/firebase.js [hunk 4]
```diff
@@ -6,6 +6,7 @@ import {
 	doc,
 	onSnapshot,
 	updateDoc,
+	increment,
```
Comment (mentalcaries) [inline src/api/firebase.js:9]: Good find with this `increment` function

File: src/components/ListItem.jsx [hunk 5]
```diff
@@ -1,5 +1,69 @@
...
+	const { name, dateLastPurchased } = item;
+
+	// Get the initial checked state from localStorage or default to false
+	const [checked, setChecked] = useState(() => {
+		const storedChecked = localStorage.getItem(`checked-${name}`);
+		return storedChecked ? JSON.parse(storedChecked) : false;
+	});
+
+	// Function to check if 24 hours have passed
+	const has24HoursPassed = (dateLastPurchased) => {
```
Comment (mentalcaries) [inline src/components/ListItem.jsx:17]: Nice function naming! I'm always a fan of easily readable code!

**Review by mentalcaries — COMMENTED** [review]
Looks good so far, but a couple of bugs to fix

File: src/components/ListItem.jsx (hunk 2 above)
Comment (arandel1) [reply src/components/ListItem.jsx:47]: **Possible solution:** totalPurchases increments by 1 when the item unchecks itself.

**Review by EmmaBin — COMMENTED** [review]
Nice work on this challenging ticket! I appreciate how readable the variable names are. After testing, it met all the criteria. Just have one small nitpick, like Devin mentioned, the totalPurchase updates every time I check and uncheck, even if it's incidental. I'm curious to learn about what is a better way to capture only valid selections—perhaps adding a confirm button after each item?

File: src/components/ListItem.jsx (hunk 2 above)
Comment (mentalcaries) [reply src/components/ListItem.jsx:49]: @arandel1 from the AC (and at least for now), the box once checked, should probably stay checked for 24 hours, as in the user won't be able to uncheck it.

Perhaps we can create another ticket for the case of accidentally checking off an item.

File: src/api/firebase.js [hunk 6]
```diff
@@ -183,12 +184,22 @@ export async function addItem(listPath, { itemName, daysUntilNextPurchase }) {
...
 }
 
-export async function updateItem() {
-	/**
-	 * TODO: Fill this out so that it uses the correct Firestore function
-	 * to update an existing item. You'll need to figure out what arguments
-	 * this function must accept!
-	 */
+// User can update items in their list by checking the checkbox associated with an item, and the totalPurchase field increments by 1 in Firebase.
+// The checkbox unchecks 24 hours after the box is checked.
```
Comment (mindyzwan) [inline src/api/firebase.js:188]: **nitpick ( non-blocking ):** Technically, this second sentence isn't relevant to this piece of code.

File: src/components/ListItem.jsx (hunk 5 above)
Comment (mindyzwan) [reply src/components/ListItem.jsx:14]: Agree!

File: src/components/ListItem.jsx [hunk 7]
```diff
@@ -1,5 +1,66 @@
...
-export function ListItem({ name }) {
-	return <li className="ListItem">{name}</li>;
+export function ListItem({ item }) {
+	// Destructure item props
+	const { name, dateLastPurchased } = item;
+
+	// Keeping [checked, setChecked] state, but removing any calls to local storage
+	const [checked, setChecked] = useState(false);
+
+	// Function to check if 24 hours have passed: Changed variable tagging.
```
Comment (mindyzwan) [inline src/components/ListItem.jsx:13]: **nitpick ( non-blocking ):** These comments are more notes to yourself than something that will likely be helpful in the long run.

File: src/utils/dates.js [hunk 8]
```diff
@@ -10,3 +10,8 @@ const ONE_DAY_IN_MILLISECONDS = 86400000;
 export function getFutureDate(offset) {
 	return new Date(Date.now() + offset * ONE_DAY_IN_MILLISECONDS);
 }
+
+// getDaysBetweenDates: We need to write a function that calculates current day - dateLastPurchased to calculate the number of days between the current day and the date last purchased
+
+//  if dateLastPurchased = null, then dateNextPurchased = null
+// dateNextPurchased = days until you need this item again + dateLastPurchased
```
Comment (mindyzwan) [inline src/utils/dates.js:17]: **nitpick ( non-blocking ):** These also look like they may have been notes left over :)

File: package.json [hunk 9]
```diff
@@ -7,6 +7,9 @@
 		"npm": ">=8.19.0"
 	},
 	"dependencies": {
+		"@emotion/react": "^11.13.3",
+		"@emotion/styled": "^11.13.0",
+		"@mui/material": "^6.0.2",
```
Comment (mindyzwan) [inline package.json:12]: **note (non-blocking):** Normally, I'd say to be sure to remove packages you decide not to use with `npm uninstall`, but I know we're using the same package in the other branch so it's all good 👍

**Review by mindyzwan — APPROVED** [review]
Looks great! Applauding y'alls tenacity on this one! 👏 Woohoo!

File: src/components/ListItem.jsx (hunk 7 above)
Comment (arandel1) [reply src/components/ListItem.jsx:13]: Removed notes

File: src/utils/dates.js (hunk 8 above)
Comment (arandel1) [reply src/utils/dates.js:17]: Removed notes

**Pull request merged**